src/xrootiface.py

```diff
@@ -86,119 +103,129 @@ def init(inconfig, inlog):
-  filename = _getfilename(filename)
+
+def stat(endpoint, filepath, userid):
+  '''Stat a file via xroot on behalf of the given userid, and returns (size, mtime). Uses the default xroot API.'''
+  filepath = _getfilepath(filepath)
   tstart = time.clock()
-  rc, statInfo = _getxrdfor(endpoint).stat(filename + _eosargs(ruid, rgid))
+  rc, statInfo = _getxrdfor(endpoint).stat(filepath + _eosargs(userid))
   tend = time.clock()
-  log.info('msg="Invoked stat" filename="%s" elapsedTimems="%.1f"' % (filename, (tend-tstart)*1000))
+  log.info('msg="Invoked stat" filepath="%s" elapsedTimems="%.1f"' % (filepath, (tend-tstart)*1000))
   if statInfo is None:
     raise IOError(rc.message.strip('\n'))
   return {'size': statInfo.size, 'mtime': statInfo.modtime}
 
-def statx(endpoint, filename, ruid, rgid):
-  '''Get extended stat info (inode, ouid, ogid, size, mtime) via an xroot opaque query on behalf of the given uid,gid'''
-  filename = _getfilename(filename)
+
+def statx(endpoint, filepath, userid):
+  '''Get extended stat info (inode, filepath, userid, size, mtime) via an xroot opaque query on behalf of the given userid'''
   tstart = time.clock()
-  rc, rawinfo = _getxrdfor(endpoint).query(QueryCode.OPAQUEFILE, filename + _eosargs(ruid, rgid) + '&mgm.pcmd=stat')
+  rc, info = _getxrdfor(endpoint).query(QueryCode.OPAQUEFILE, _getfilepath(filepath) + _eosargs(userid) + '&mgm.pcmd=stat')
   tend = time.clock()
-  log.info('msg="Invoked stat" filename="%s" elapsedTimems="%.1f"' % (filename, (tend-tstart)*1000))
+  log.info('msg="Invoked stat" filepath="%s" elapsedTimems="%.1f"' % (_getfilepath(filepath), (tend-tstart)*1000))
   if '[SUCCESS]' not in str(rc):
     raise IOError(str(rc).strip('\n'))
-  rawinfo = str(rawinfo)
-  if 'retc=' in rawinfo:
-    raise IOError(rawinfo.strip('\n'))
-  statxdata = rawinfo.split()
-  return {'inode': statxdata[2],
-          'ouid': statxdata[5],
-          'ogid': statxdata[6],
+  info = str(info)
+  if 'retc=' in info:
+    raise IOError(info.strip('\n'))
+  statxdata = info.split()
+  return {'inode': str(statxdata[2]),
+          'filepath': filepath,
+          'userid': str(statxdata[5]) + ':' + str(statxdata[6]),
           'size': int(statxdata[8]),
           'mtime': statxdata[12]}
 
-def setxattr(endpoint, filename, ruid, rgid, key, value):
-  '''Set the extended attribute <key> to <value> via a special open on behalf of the given uid, gid'''
-  _xrootcmd(endpoint, 'attr', 'set', ruid, rgid, 'mgm.attr.key=' + key + '&mgm.attr.value=' + str(value) + '&mgm.path=' + _getfilename(filename))
 
-def getxattr(endpoint, filename, ruid, rgid, key):
-  '''Get the extended attribute <key> via a special open on behalf of the given uid, gid'''
-  res = _xrootcmd(endpoint, 'attr', 'get', ruid, rgid, 'mgm.attr.key=' + key + '&mgm.path=' + _getfilename(filename))
+def setxattr(endpoint, filepath, userid, key, value):
+  '''Set the extended attribute <key> to <value> via a special open on behalf of the given userid'''
```

Comment:
`..via a special open on behalf..`?

Reply:
If you look at the real implementation in `_xrootcmd`, you have a clue why I call it a "special open": to perform those metadata operations, we have to open() "special" files in a kind of system path, which is then interpreted by eos as the command on the real target file.
It's actually a pretty ugly interface, but that's it ... the xroot python bindings do not provide the full API we need, as the CS3 APIs now do.

Reply:
Maybe we can just improve the docs here, so one doesn't have to look somewhere else?

Reply:
Well, one *has* to look at the eos internals to understand how to perform those actions. Possibly by now these have been documented, definitely not at the time.
Anyway this PR didn't really touch `xrootiface.py`, the only change is the (uid, gid) tuple got replaced by userid, so I would not further improve the xroot layer in this context.